File: language/runtime/apple/ios/Sources/RuntimeHostIOS/Module/Background/BackgroundRegistrationStore.swift
import Foundation
import RuntimeHostAppleCore

private let backgroundRegistrationsKey = "dev.destack.runtime.ios.background.registrations"

/// One durable iOS background registration record.
internal struct BackgroundTaskRecord: Codable {
  /// The stored task descriptor.
  let descriptor: RuntimeHostBackgroundTaskDescriptor

  /// The next regular recurring execution target when this task recurs.
  let nextRegularRunUnixNs: UInt64?
}

/// One background registration storage result.
internal struct BackgroundStoredRegistration {
  /// The active record for this identifier.
  let record: BackgroundTaskRecord

  /// Whether the scheduler should update this registration.
  let shouldSchedule: Bool
}

/// The registration store for one iOS background host.
@MainActor
internal final class BackgroundRegistrationStore {
  /// The user defaults backing store.
  private let defaults: UserDefaults

  /// The identifiers with registered launch handlers.
  private(set) var registeredIdentifiers: Set<String>

  /// The registered records by stable identifier.
  private(set) var registrations: [String: BackgroundTaskRecord]

  /// Create one background registration store.
  init(
    defaults: UserDefaults
  ) {
    // capture the persistent store and restore the durable registrations
    self.defaults = defaults
    self.registrations = Self.loadRegistrations(defaults)
    self.registeredIdentifiers = []
  }

  /// Return every descriptor sorted by stable identifier.
  func listDescriptors() -> [RuntimeHostBackgroundTaskDescriptor] {
    // return one stable listing sorted by identifier
    registrations
      .values
      .map(\.descriptor)
      .sorted(by: { $0.identifier < $1.identifier })
  }

  /// Resolve one record by stable identifier.
  func resolveRecord(
    _ identifier: String
  ) -> BackgroundTaskRecord? {
    registrations[identifier]
  }

  /// Resolve one descriptor by stable identifier.
  func resolveDescriptor(
    _ identifier: String
  ) -> RuntimeHostBackgroundTaskDescriptor? {
    registrations[identifier]?.descriptor
  }

  /// Persist one registration descriptor.
  func putRegistration(
    _ descriptor: RuntimeHostBackgroundTaskDescriptor,
    nextRegularRunUnixNs: UInt64?
  ) -> BackgroundStoredRegistration {
    // keep one existing registration when the conflict policy says so
    if let existingRecord = registrations[descriptor.identifier],
      descriptor.conflictPolicy == .keep
    {
      return BackgroundStoredRegistration(
        record: existingRecord,
        shouldSchedule: false
      )
    }

    let record = BackgroundTaskRecord(
      descriptor: descriptor,
      nextRegularRunUnixNs: nextRegularRunUnixNs
    )

    // store the registration under its stable identifier
    registrations[descriptor.identifier] = record

    // flush the durable registration snapshot
    storeRegistrations()

    return BackgroundStoredRegistration(
      record: record,
      shouldSchedule: true
    )
  }

  /// Persist one full registration record.
  func putRecord(
    _ record: BackgroundTaskRecord
  ) {
    registrations[record.descriptor.identifier] = record

    storeRegistrations()
  }

  /// Remove one registration descriptor.
  func removeRegistration(
    _ identifier: String
  ) -> BackgroundTaskRecord? {
    // remove the registration from the live map
    let record = registrations.removeValue(forKey: identifier)

    // flush the durable registration snapshot when it changed
    if record != nil {
      storeRegistrations()
    }

    return record
  }

  /// Return whether the launch handler for this identifier is already registered.
  func isHandlerRegistered(
    _ identifier: String
  ) -> Bool {
    registeredIdentifiers.contains(identifier)
  }

  /// Mark one handler identifier as registered.
  func markHandlerRegistered(
    _ identifier: String
  ) {
    registeredIdentifiers.insert(identifier)
  }

  /// Load every durable registration descriptor from user defaults.
  private static func loadRegistrations(
    _ defaults: UserDefaults
  ) -> [String: BackgroundTaskRecord] {
    // load the stored payload when one exists
    guard let data = defaults.data(forKey: backgroundRegistrationsKey) else {
      return [:]
    }

    // decode the payload into runtime descriptors
    guard
      let records = try? JSONDecoder().decode(
        [BackgroundTaskRecord].self,
        from: data
      )
    else {
      return [:]
    }

    // index the descriptors by stable identifier
    return Dictionary(
      uniqueKeysWithValues: records.map { record in
        (record.descriptor.identifier, record)
      }
    )
  }

  /// Persist every durable registration descriptor into user defaults.
  private func storeRegistrations() {
    // keep every durable registration in a stable order
    let descriptors =
      registrations
      .values
      .sorted(by: { left, right in left.descriptor.identifier < right.descriptor.identifier })

    // encode the durable registration snapshot
    guard let data = try? JSONEncoder().encode(descriptors) else {
      return
    }

    // write the snapshot into user defaults
    defaults.set(data, forKey: backgroundRegistrationsKey)
  }
}

File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/background/BackgroundRegistrationStore.kt
package dev.destack.runtime.android.module.background

import android.content.SharedPreferences

import org.json.JSONArray
import org.json.JSONObject

internal const val backgroundPreferencesName: String = "dev.destack.runtime.background"

private const val backgroundRecordsKey: String = "records"

/**
 * One durable Android background task record.
 */
internal data class BackgroundTaskRecord(
    /**
     * The stored task descriptor.
     */
    val descriptor: RuntimeHostBackgroundTaskDescriptor,

    /**
     * The next regular recurring execution target when this task recurs.
     */
    val nextRegularRunUnixNs: Long?,
)

/**
 * One background registration storage result.
 */
internal data class BackgroundStoredRegistration(
    /**
     * The active record for the identifier.
     */
    val record: BackgroundTaskRecord,

    /**
     * Whether the scheduler should update this registration.
     */
    val shouldSchedule: Boolean,
)

/**
 * The task registration store for one Android process.
 */
internal class BackgroundRegistrationStore(
    private val preferences: SharedPreferences,
) {
    /**
     * The durable records loaded from shared preferences.
     */
    private val records: MutableMap<String, BackgroundTaskRecord> =
        loadRecords(preferences)

    /**
     * Return every registered task descriptor.
     */
    fun listDescriptors(): List<RuntimeHostBackgroundTaskDescriptor> {
        return records
            .values
            .map { record -> record.descriptor }
            .sortedBy { descriptor -> descriptor.identifier }
    }

    /**
     * Resolve one record by stable identifier.
     */
    fun resolveRecord(
        identifier: String,
    ): BackgroundTaskRecord? {
        return records[identifier]
    }

    /**
     * Store one registration payload and report whether the scheduler should update it.
     */
    fun putRegistration(
        options: RuntimeHostBackgroundTaskOptions,
        nextRegularRunUnixNs: Long?,
    ): BackgroundStoredRegistration {
        val existingRecord = records[options.identifier]
        if (
            existingRecord != null &&
            options.conflictPolicy == RuntimeHostBackgroundConflictPolicy.Keep
        ) {
            return BackgroundStoredRegistration(
                record = existingRecord,
                shouldSchedule = false,
            )
        }

        val record = BackgroundTaskRecord(
            descriptor = descriptorFromOptions(options),
            nextRegularRunUnixNs = nextRegularRunUnixNs,
        )

        records[options.identifier] = record

        storeRecords(preferences, records.values)

        return BackgroundStoredRegistration(
            record = record,
            shouldSchedule = true,
        )
    }

    /**
     * Remove one registration by stable identifier.
     */
    fun removeRegistration(
        identifier: String,
    ): BackgroundTaskRecord? {
        val record = records.remove(identifier)

        if (record != null) {
            storeRecords(preferences, records.values)
        }

        return record
    }

    /**
     * Replace one stored record under its stable identifier.
     */
    fun putRecord(
        record: BackgroundTaskRecord,
    ) {
        records[record.descriptor.identifier] = record

        storeRecords(preferences, records.values)
    }
}

/**
 * Build one task descriptor view from one registration payload.
 */
private fun descriptorFromOptions(
    options: RuntimeHostBackgroundTaskOptions,
): RuntimeHostBackgroundTaskDescriptor {
    return RuntimeHostBackgroundTaskDescriptor(
        identifier = options.identifier,
        trigger = options.trigger,
        schedule = options.schedule,
        network = options.network,
        requiresCharging = options.requiresCharging,
        requiresIdle = options.requiresIdle,
        conflictPolicy = options.conflictPolicy,
    )
}

/**
 * Load every durable background record from shared preferences.
 */
private fun loadRecords(
    preferences: SharedPreferences,
): MutableMap<String, BackgroundTaskRecord> {
    val payload = preferences.getString(backgroundRecordsKey, null)
        ?: return mutableMapOf()
    val entries = JSONArray(payload)
    val records = mutableMapOf<String, BackgroundTaskRecord>()

    for (index in 0 until entries.length()) {
        val entry = entries.getJSONObject(index)
        val descriptor = RuntimeHostBackgroundTaskDescriptor(
            identifier = entry.getString("identifier"),
            trigger = decodeTrigger(entry.getInt("trigger")) ?: continue,
            schedule = RuntimeHostBackgroundTaskSchedule(
                kind = decodeScheduleKind(entry.getInt("scheduleKind")) ?: continue,
                earliestBeginUnixNs =
                    if (entry.isNull("earliestBeginUnixNs")) null else entry.getLong(
                        "earliestBeginUnixNs",
                    ),
                repeatIntervalNs =
                    if (entry.isNull("repeatIntervalNs")) null else entry.getLong(
                        "repeatIntervalNs",
                    ),
            ),
            network = decodeNetwork(entry.getInt("network")) ?: continue,
            requiresCharging = entry.getBoolean("requiresCharging"),
            requiresIdle = entry.getBoolean("requiresIdle"),
            conflictPolicy = decodeConflictPolicy(entry.getInt("conflictPolicy")) ?: continue,
        )
        val record = BackgroundTaskRecord(
            descriptor = descriptor,
            nextRegularRunUnixNs =
                if (entry.isNull("nextRegularRunUnixNs")) null else entry.getLong(
                    "nextRegularRunUnixNs",
                ),
        )

        records[descriptor.identifier] = record
    }

    return records
}

/**
 * Persist every durable background record into shared preferences.
 */
private fun storeRecords(
    preferences: SharedPreferences,
    records: Collection<BackgroundTaskRecord>,
) {
    val payload = JSONArray()

    for (record in records.sortedBy { value -> value.descriptor.identifier }) {
        val descriptor = record.descriptor

        payload.put(
            JSONObject()
                .put("identifier", descriptor.identifier)
                .put("trigger", encodeTrigger(descriptor.trigger))
                .put("scheduleKind", encodeScheduleKind(descriptor.schedule.kind))
                .put("earliestBeginUnixNs", descriptor.schedule.earliestBeginUnixNs)
                .put("repeatIntervalNs", descriptor.schedule.repeatIntervalNs)
                .put("network", encodeNetwork(descriptor.network))
                .put("requiresCharging", descriptor.requiresCharging)
                .put("requiresIdle", descriptor.requiresIdle)
                .put("conflictPolicy", encodeConflictPolicy(descriptor.conflictPolicy))
                .put("nextRegularRunUnixNs", record.nextRegularRunUnixNs),
        )
    }

    preferences.edit()
        .putString(backgroundRecordsKey, payload.toString())
        .apply()
}

/**
 * Encode one runtime trigger kind as one stored integer tag.
 */
private fun encodeTrigger(
    trigger: RuntimeHostBackgroundTriggerKind,
): Int {
    return when (trigger) {
        RuntimeHostBackgroundTriggerKind.AppRefresh -> 1
        RuntimeHostBackgroundTriggerKind.Processing -> 2
    }
}

/**
 * Decode one stored integer tag as one runtime trigger kind.
 */
private fun decodeTrigger(
    value: Int,
): RuntimeHostBackgroundTriggerKind? {
    return when (value) {
        1 -> RuntimeHostBackgroundTriggerKind.AppRefresh
        2 -> RuntimeHostBackgroundTriggerKind.Processing
        else -> null
    }
}

/**
 * Encode one runtime network requirement as one stored integer tag.
 */
private fun encodeNetwork(
    network: RuntimeHostBackgroundNetworkRequirement,
): Int {
    return when (network) {
        RuntimeHostBackgroundNetworkRequirement.None -> 1
        RuntimeHostBackgroundNetworkRequirement.Connected -> 2
        RuntimeHostBackgroundNetworkRequirement.Unmetered -> 3
    }
}

/**
 * Encode one runtime schedule kind as one stored integer tag.
 */
private fun encodeScheduleKind(
    kind: RuntimeHostBackgroundTaskScheduleKind,
): Int {
    return when (kind) {
        RuntimeHostBackgroundTaskScheduleKind.Once -> 1
        RuntimeHostBackgroundTaskScheduleKind.Recurring -> 2
    }
}

/**
 * Decode one stored integer tag as one runtime schedule kind.
 */
private fun decodeScheduleKind(
    value: Int,
): RuntimeHostBackgroundTaskScheduleKind? {
    return when (value) {
        1 -> RuntimeHostBackgroundTaskScheduleKind.Once
        2 -> RuntimeHostBackgroundTaskScheduleKind.Recurring
        else -> null
    }
}

/**
 * Decode one stored integer tag as one runtime network requirement.
 */
private fun decodeNetwork(
    value: Int,
): RuntimeHostBackgroundNetworkRequirement? {
    return when (value) {
        1 -> RuntimeHostBackgroundNetworkRequirement.None
        2 -> RuntimeHostBackgroundNetworkRequirement.Connected
        3 -> RuntimeHostBackgroundNetworkRequirement.Unmetered
        else -> null
    }
}

/**
 * Encode one runtime conflict policy as one stored integer tag.
 */
private fun encodeConflictPolicy(
    policy: RuntimeHostBackgroundConflictPolicy,
): Int {
    return when (policy) {
        RuntimeHostBackgroundConflictPolicy.Replace -> 1
        RuntimeHostBackgroundConflictPolicy.Keep -> 2
    }
}

/**
 * Decode one stored integer tag as one runtime conflict policy.
 */
private fun decodeConflictPolicy(
    value: Int,
): RuntimeHostBackgroundConflictPolicy? {
    return when (value) {
        1 -> RuntimeHostBackgroundConflictPolicy.Replace
        2 -> RuntimeHostBackgroundConflictPolicy.Keep
        else -> null
    }
}
